metrics/api/src/main/java/io/helidon/metrics/api/Timer.java

```diff
@@ -0,0 +1,178 @@
+     *
+     * @param f runnable to be timed
+     */
+    void record(Runnable f);
+
+    /**
+     * Wraps a {@link Runnable} so that it will be timed every time it is invoked via the return value from this method.
+     *
+     * @param f runnable to time when it is invoked
+     * @return the wrapped runnable
+     */
+    Runnable wrap(Runnable f);
+
+    /**
+     * Wraps a {@link Callable} so that it is will be timed every time it is invoked via the return value from this method.
+     *
+     * @param f   callable to time when it is invoked
+     * @param <T> return type of the callable
+     * @return the wrapped callable
+     */
+    <T> Callable<T> wrap(Callable<T> f);
+
+    /**
+     * Wraps a {@link Supplier} so that it will be timed every time it is invoked via the return value from this method.
+     *
+     * @param f   {@code Supplier} to time when it is invoked
+     * @param <T> return type of the {@code Supplier} result
+     * @return the wrapped supplier
+     */
+    <T> Supplier<T> wrap(Supplier<T> f);
+
+    /**
+     * Returns the current count of completed events measured by the timer.
+     *
+     * @return number of events recorded by the timer
+     */
+    long count();
+
+    /**
+     * Returns the total time, expressed in the specified units, consumed by completed events
+     * measured by the timer.
+     *
+     * @param unit time unit in which to express the total accumulated time
+     * @return total time of recorded events
+     */
+    double totalTime(TimeUnit unit);
+
+    /**
+     * Returns the average time, expressed in the specified units, consumed by completed events
+     * measured by the timer.
+     *
+     * @param unit time unit in which to express the mean
+     * @return average for all events recorded by this timer
+     */
+    double mean(TimeUnit unit);
+
+    /**
+     * Returns the maximum value, expressed in the specified units, consumed by a completed event
+     * measured by the timer.
+     *
+     * @param unit time unit in which to express the maximum
+     * @return maximum time recorded by a single event measured by this timer
+     */
+    double max(TimeUnit unit);
+
+    /**
+     * Measures an interval of time from instantiation to an explicit invocation of {@link #stop(io.helidon.metrics.api.Timer)}.
+     * <p>
+     *     A {@code Sample} is not bound to a specific {@code Timer} until it is stopped, at
+     *     which time the caller specifies which timer to update.
+     * </p>
+     */
+    interface Sample {
+
+        /**
+         * Ends the interval, recording the current time as the end time of the interval and applying the elapsed time to the
+         * specified {@link io.helidon.metrics.api.Timer}.
+         *
+         * @param timer the timer to update with this interval
+         * @return duration of the sample (in nanoseconds)
+         */
+        long stop(Timer timer);
```

Comment:
We should not need the `Timer` parameter, as the `Sample` is created by the timer itself, so we should be able to keep track of it within the Sample instance.

Reply:
Note that as proposed (following the Micrometer pattern) the `Sample` is returned from a _static_ method on `Timer`. The binding of a `Sample` to a `Timer` is deferred until the moment when the sample is stopped. 

Conceptually this makes sense: there is nothing inherent in the current state of a timer that influences the creation of a sample.

Further, there could be use cases in which the specific timer we want to update depends on information known only when the sample stops. For example, we might have separate timers for successful and unsuccessful HTTP request processing. We cannot know which timer we will want to update when we create the `Sample`, only when we stop it.

Reply:
Agreed, thanks for the explanation